library/alloc/src/vec/mod.rs

```diff
@@ -1514,15 +1514,102 @@ impl<T, A: Allocator> Vec<T, A> {
     /// assert_eq!(vec, ["foo", "bar", "baz", "bar"]);
     /// ```
     #[stable(feature = "dedup_by", since = "1.16.0")]
-    pub fn dedup_by<F>(&mut self, same_bucket: F)
+    pub fn dedup_by<F>(&mut self, mut same_bucket: F)
     where
         F: FnMut(&mut T, &mut T) -> bool,
     {
-        let len = {
-            let (dedup, _) = self.as_mut_slice().partition_dedup_by(same_bucket);
-            dedup.len()
-        };
-        self.truncate(len);
+        let len = self.len();
+        if len <= 1 {
+            return;
+        }
+
+        /* INVARIANT: vec.len() > read >= write > write-1 >= 0 */
```

Comment:
Nit: isn't this equivalent to `vec.len() > read > write-1 >= 0`? I feel phrasing the invariant this way makes it significantly more obvious what it is about.

Reply:
Later in code it matters, because `read` can be equal to `write` making `read_ptr == write_ptr`